File: com-mini-mvc-test/src/main/java/com/mini/web/test/dao/UserDao.java
package com.mini.web.test.dao;

import com.google.inject.ImplementedBy;
import com.mini.jdbc.SQLBuilder;
import com.mini.jdbc.util.Paging;
import com.mini.util.StringUtil;
import com.mini.web.test.dao.base.BaseUserDao;
import com.mini.web.test.dao.impl.UserDaoImpl;
import com.mini.web.test.entity.Region;
import com.mini.web.test.entity.extend.UserExt;
import com.mini.web.test.entity.extend.UserExt.UserExtendBuilder;

import java.time.LocalDate;
import java.util.List;

import static com.mini.util.DateUtil.formatDate;
import static com.mini.web.test.entity.User.*;

/**
 * UserDao.java
 * @author xchao
 */
@ImplementedBy(UserDaoImpl.class)
public interface UserDao extends BaseUserDao {

    default List<UserExt> search(Paging paging, String search, int sortType, int phoneAuto, int emailAuto, String regionIdUri,
            LocalDate startTime, LocalDate endTime) {
        return query(paging, new UserExtendBuilder() {
            {
                // 搜索关键字条件
                if (!StringUtil.isBlank(search)) {
                    where("(%s LIKE ? OR %s LIKE ? OR %s LIKE ?)", NAME, FULL_NAME, PHONE);
                    params(search + "%", search + "%", search + "%");
                }

                // 手机号已认证
                if (phoneAuto == 1) {
                    where("%s = ?", PHONE_AUTH);
                    params(phoneAuto);
                }

                // 油箱已认证
                if (emailAuto == 1) {
                    where("%s = ?", EMAIL_AUTH);
                    params(emailAuto);
                }

                // 地区条件
                if (!StringUtil.isBlank(regionIdUri)) {
                    where("%s LIKE ?", Region.ID_URI);
                    params(regionIdUri + "%");
                }

                // 开始时间
                if (startTime != null) {
                    where("%s >= ? ", CREATE_TIME);
                    params(formatDate(startTime));
                }

                // 结束时间
                if (endTime != null) {
                    where("%s < ? ", CREATE_TIME);
                    params(formatDate(endTime.plusDays(1)));
                }
                // 排序
                order_by("%s %s", CREATE_TIME, (sortType == 1 ? "DESC" : "ASC"));
            }
        }, UserExt::mapper);
    }

    default int delete(long[] idList) {
        return execute(new SQLBuilder() {
            {
                delete().from(TABLE);
                for (long id : idList) {
                    or();
                    where("%s = ?", ID);
                    params(id);
                }
            }
        });
    }
}

File: com-mini-mvc-kotlin/src/main/kotlin/com/mini/mvc/kotlin/dao/UserDao.kt
package com.mini.mvc.kotlin.dao

import com.mini.jdbc.BasicsDao
import com.mini.jdbc.SQLBuilder
import com.mini.jdbc.util.Paging
import com.mini.mvc.kotlin.entity.User
import com.mini.mvc.kotlin.entity.mapper.UserMapper
import com.mini.mvc.kotlin.entity.mapper.UserMapper.UserBuilder

interface UserDao : BasicsDao {
    var userMapper: UserMapper?

    /**
     * 添加实体信息
     * @param user 实体信息
     * @return 执行结果
     */
    @Suppress("unused")
    fun insert(user: User): Int {
        return execute(object : SQLBuilder() {
            init {
                insert_into(User.TABLE)
                // 用户ID
                values(User.ID)
                params(user.id)
                // 用户名
                values(User.NAME)
                params(user.name)
                // MD5(密码)
                values(User.PASSWORD)
                params(user.password)
                // 用户手机号
                values(User.PHONE)
                params(user.phone)
                // 0-未认证，1-已谁
                values(User.PHONE_AUTH)
                params(user.phoneAuth)
                // 用户姓名
                values(User.FULL_NAME)
                params(user.fullName)
                // 用户邮箱地址
                values(User.EMAIL)
                params(user.email)
                // 0-未认证，1-已认证
                values(User.EMAIL_AUTH)
                params(user.emailAuth)
                // 用户头像地址
                values(User.HEAD_URL)
                params(user.headUrl)
                // 用户所属地区ID
                values(User.REGION_ID)
                params(user.regionId)
                // 用户注册时间
                values(User.CREATE_TIME)
                params(user.createTime)
            }
        })
    }

    /**
     * 添加实体信息
     * @param user 实体信息
     * @return 执行结果
     */
    @Suppress("unused")
    fun replace(user: User): Int {
        return execute(object : SQLBuilder() {
            init {
                replace_into(User.TABLE)
                // 用户ID
                values(User.ID)
                params(user.id)
                // 用户名
                values(User.NAME)
                params(user.name)
                // MD5(密码)
                values(User.PASSWORD)
                params(user.password)
                // 用户手机号
                values(User.PHONE)
                params(user.phone)
                // 0-未认证，1-已谁
                values(User.PHONE_AUTH)
                params(user.phoneAuth)
                // 用户姓名
                values(User.FULL_NAME)
                params(user.fullName)
                // 用户邮箱地址
                values(User.EMAIL)
                params(user.email)
                // 0-未认证，1-已认证
                values(User.EMAIL_AUTH)
                params(user.emailAuth)
                // 用户头像地址
                values(User.HEAD_URL)
                params(user.headUrl)
                // 用户所属地区ID
                values(User.REGION_ID)
                params(user.regionId)
                // 用户注册时间
                values(User.CREATE_TIME)
                params(user.createTime)
            }
        })
    }

    /**
     * 修改实体信息
     * @param user 实体信息
     * @return 执行结果
     */
    @Suppress("unused")
    fun update(user: User): Int {
        return execute(object : SQLBuilder() {
            init {
                update(User.TABLE)
                // 用户ID
                set("%s = ?", User.ID)
                params(user.id)
                // 用户名
                set("%s = ?", User.NAME)
                params(user.name)
                // MD5(密码)
                set("%s = ?", User.PASSWORD)
                params(user.password)
                // 用户手机号
                set("%s = ?", User.PHONE)
                params(user.phone)
                // 0-未认证，1-已谁
                set("%s = ?", User.PHONE_AUTH)
                params(user.phoneAuth)
                // 用户姓名
                set("%s = ?", User.FULL_NAME)
                params(user.fullName)
                // 用户邮箱地址
                set("%s = ?", User.EMAIL)
                params(user.email)
                // 0-未认证，1-已认证
                set("%s = ?", User.EMAIL_AUTH)
                params(user.emailAuth)
                // 用户头像地址
                set("%s = ?", User.HEAD_URL)
                params(user.headUrl)
                // 用户所属地区ID
                set("%s = ?", User.REGION_ID)
                params(user.regionId)
                // 用户注册时间
                set("%s = ?", User.CREATE_TIME)
                params(user.createTime)
                // 用户ID
                where("%s = ?", User.ID)
                params(user.id)
            }
        })
    }

    /**
     * 删除实体信息
     * @param user 实体信息
     * @return 执行结果
     */
    fun delete(user: User): Int {
        return execute(object : SQLBuilder() {
            init {
                delete().from(User.TABLE)
                // 用户ID
                where("%s = ?", User.ID)
                params(user.id)
            }
        })
    }

    /**
     * 根据ID删除实体信息
     * @param id 用户ID
     * @return 执行结果
     */
    @Suppress("unused")
    fun deleteById(id: Long): Int {
        return execute(object : SQLBuilder() {
            init {
                delete().from(User.TABLE)
                // 用户ID
                where("%s = ?", User.ID)
                params(id)
            }
        })
    }

    /**
     * 根据ID查询实体信息
     * @param id 用户ID
     * @return 实体信息
     */
    @Suppress("unused")
    fun queryById(id: Long): User {
        return queryOne<User>(object : UserBuilder() {
            init {
                // 用户ID
                where("%s = ?", User.ID)
                params(id)
            }
        }, userMapper)
    }

    /**
     * 查询所有实体信息
     * @return 实体信息列表
     */
    @Suppress("unused")
    fun queryAll(): List<User> {
        return query<User>(object : UserBuilder() {
            init {
                //
            }
        }, userMapper)
    }

    /**
     * 查询所有实体信息
     * @param paging 分布工具
     * @return 实体信息列表
     */
    @Suppress("unused")
    fun queryAll(paging: Paging): List<User> {
        return query<User>(paging, object : UserBuilder() {
            init {
                //
            }
        }, userMapper)
    }
}